guardrails_test.py

```diff
@@ -0,0 +1,96 @@
+import logging
+import time
+import pytest
+
+from cassandra import InvalidRequest
+
+from dtest import Tester, create_ks
+from tools.assertions import assert_one
+
+since = pytest.mark.since
+logger = logging.getLogger(__name__)
+
+class BaseGuardrailsTester(Tester):
+
+    def prepare(self, rf=1, options=None, nodes=3, install_byteman=False, extra_jvm_args=None, **kwargs):
+        if options is None:
+            options = {}
+
+        if extra_jvm_args is None:
+            extra_jvm_args = []
+
+        cluster = self.cluster
+        cluster.populate([nodes, 0], install_byteman=install_byteman)
+        if options:
+            cluster.set_configuration_options(values=options)
+
+        cluster.start(jvm_args=extra_jvm_args)
+        node1 = cluster.nodelist()[0]
+
+        session = self.patient_cql_connection(node1, **kwargs)
+        create_ks(session, 'ks', rf)
+
+        return session
+
+
+@since('4.0')
+class TestGuardrails(BaseGuardrailsTester):
+
+    def test_disk_usage_guardrail(self):
+        """
+        Test disk usage guardrail will warn if exceeds warn threshold and reject writes if exceeds failure threshold
+        """
+
+        self.fixture_dtest_setup.ignore_log_patterns = ["Write request failed because disk usage exceeds failure threshold"]
+        guardrails_config = {'guardrails': {'disk_usage_percentage_warn_threshold': 98,
+                                            'disk_usage_percentage_failure_threshold': 99}}
+
+        logger.debug("prepare 2-node cluster with rf=1 and guardrails enabled")
+        session = self.prepare(rf=1, nodes=2, options=guardrails_config, extra_jvm_args=['-Dcassandra.disk_usage.monitor_interval_ms=100'], install_byteman=True)
+        node1, node2 = self.cluster.nodelist()
+        session.execute("CREATE TABLE t (id int PRIMARY KEY, v int)")
+
+        logger.debug("Inject FULL to node1, expect log on node1 and node2 rejects writes")
+        self.disk_usage_injection(node1, "full", False)
+        time.sleep(2)  # wait for disk state broadcast
```

Comment:
disk state is broadcast in gossip?
is there a way to verify broadcast instead of sleeping?
it will bite us eventually as a flaky test

Reply:
I don't know either

Reply:
@Gerrrr maybe you know?

Reply:
> is disk state broadcast in gossip?

Yes, see [DiskUsageBroadcaster](https://github.com/riptano/bdp/blob/a0c44fb8c8f0876412b098fe921b44eadc210a71/dse-db/src/java/org/apache/cassandra/service/disk/usage/DiskUsageBroadcaster.java#L42).

>  is there a way to verify broadcast instead of sleeping?

Good idea. We can probably add a trace log around [here](https://github.com/riptano/bdp/blob/a0c44fb8c8f0876412b098fe921b44eadc210a71/dse-db/src/java/org/apache/cassandra/service/disk/usage/DiskUsageBroadcaster.java#L103) and wait for it to appear.

Reply:
Thanks for the idea!

It's interesting - I tried it and it works, but the test became flaky! Failing about 1/3 times. It looks like the message is logged before the new state is effective.

I found an existing message in the log
```
TRACE [GossipStage:1] 2021-06-04 10:20:35,486 Gossiper.java:1152 - Adding state DISK_USAGE: FULL
```
And waiting for this to appear in the log is working consistently.